src/main/java/io/jenkins/plugins/sshbuildagents/ssh/mina/SSHApacheMinaLauncher.java

```diff
@@ -0,0 +1,1125 @@
+        return getWorkingDirectory(computer.getNode());
+    }
+
+    /**
+     * Returns the remote root workspace (without trailing slash).
+     *
+     * @param agent The agent to get the root workspace of.
+     * @return the remote root workspace (without trailing slash).
+     */
+    @CheckForNull
+    private static String getWorkingDirectory(@CheckForNull Slave agent) {
+        if (agent == null) {
+            return null;
+        }
+        String workingDirectory = agent.getRemoteFS();
+        while (workingDirectory.endsWith("/")) {
+            workingDirectory = workingDirectory.substring(0, workingDirectory.length() - 1);
+        }
+        return workingDirectory;
+    }
+
+    /**
+     * Gets the credentials used to connect to the agent.
+     *
+     * @return The credentials used to connect to the agent.
+     */
+    public StandardUsernameCredentials getCredentials() {
+        String credentialsId = this.credentialsId == null
+                ? (this.credentials == null ? null : this.credentials.getId())
+                : this.credentialsId;
+        try {
+            // only ever want from the system
+            // lookup every time so that we always have the latest
+            StandardUsernameCredentials credentials =
+                    credentialsId != null ? SSHApacheMinaLauncher.lookupSystemCredentials(credentialsId) : null;
+            if (credentials != null) {
+                this.credentials = credentials;
+                return credentials;
+            }
+        } catch (Throwable t) {
+            // ignore
+        }
+
+        return this.credentials;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public boolean isLaunchSupported() {
+        return connection == null;
+    }
+
+    /**
+     * Gets the optional JVM Options used to launch the agent JVM.
+     *
+     * @return The optional JVM Options used to launch the agent JVM.
+     */
+    public String getJvmOptions() {
+        return jvmOptions == null ? "" : jvmOptions;
+    }
+
+    /** Sets the optional JVM Options used to launch the agent JVM. */
+    @DataBoundSetter
+    public void setJvmOptions(String value) {
+        this.jvmOptions = fixEmpty(value);
+    }
+
+    /**
+     * Gets the optional java command to use to launch the agent JVM.
+     *
+     * @return The optional java command to use to launch the agent JVM.
+     */
+    public String getJavaPath() {
+        return javaPath == null ? "" : javaPath;
+    }
+
+    @DataBoundSetter
+    public void setJavaPath(String value) {
+        this.javaPath = fixEmpty(value);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void launch(@NonNull final SlaveComputer computer, @NonNull final TaskListener listener)
+            throws InterruptedException {
+        final Node node = computer.getNode();
+        final String host = this.host;
+        final int port = this.port;
+        if (computer == null || listener == null) {
```

Comment:
not sure it's needed as both have `@NonNull ` annotation :)

Reply:
I struggle to remove a warning from SpotBugs; this is a leftover. I will remove it.